pkg/query/intersection_test.go

```diff
@@ -285,3 +299,220 @@ func TestIntersectionIteratorEarlyTermination(t *testing.T) {
+		require.Equal("doc1", path.Resource.ObjectID)
+		require.Equal("user", path.Subject.ObjectType)
+		require.Equal("alice", path.Subject.ObjectID)
+		// The relation should be empty since different relations were merged (per Path.mergeFrom logic)
+		require.Equal("", path.Relation, "Different relations should result in empty relation")
+	})
+
+	t.Run("No_Common_Endpoints", func(t *testing.T) {
+		t.Parallel()
+
+		// Iterators with no common endpoints - should return empty
+		pathDoc1Alice := MustPathFromString("document:doc1#viewer@user:alice")
+		pathDoc1Alice.Caveat = &core.CaveatExpression{
+			OperationOrCaveat: &core.CaveatExpression_Caveat{
+				Caveat: &core.ContextualizedCaveat{
+					CaveatName: "caveat1",
+				},
+			},
+		}
+
+		pathDoc2Bob := MustPathFromString("document:doc2#editor@user:bob")
+		pathDoc2Bob.Caveat = &core.CaveatExpression{
+			OperationOrCaveat: &core.CaveatExpression_Caveat{
+				Caveat: &core.ContextualizedCaveat{
+					CaveatName: "caveat2",
+				},
+			},
+		}
+
+		iter1 := NewFixedIterator(pathDoc1Alice)
+		iter2 := NewFixedIterator(pathDoc2Bob)
+
+		intersect := NewIntersection()
+		intersect.addSubIterator(iter1)
+		intersect.addSubIterator(iter2)
+
+		relSeq, err := ctx.Check(intersect, NewObjects("document", "doc1", "doc2"), NewObject("user", "alice").WithEllipses())
+		require.NoError(err)
+
+		rels, err := CollectAll(relSeq)
+		require.NoError(err)
+
+		require.Len(rels, 0, "Different endpoints should not intersect - should return empty")
+	})
+
+	t.Run("Three_Iterators_Mixed_Caveats", func(t *testing.T) {
+		t.Parallel()
+
+		// Three iterators with the same path but different caveat combinations
+		pathCaveat1 := MustPathFromString("document:doc1#viewer@user:alice")
+		pathCaveat1.Caveat = &core.CaveatExpression{
+			OperationOrCaveat: &core.CaveatExpression_Caveat{
+				Caveat: &core.ContextualizedCaveat{
+					CaveatName: "caveat1",
+				},
+			},
+		}
+
+		pathNoCaveat := MustPathFromString("document:doc1#viewer@user:alice")
+
+		pathCaveat2 := MustPathFromString("document:doc1#viewer@user:alice")
+		pathCaveat2.Caveat = &core.CaveatExpression{
+			OperationOrCaveat: &core.CaveatExpression_Caveat{
+				Caveat: &core.ContextualizedCaveat{
+					CaveatName: "caveat2",
+				},
+			},
+		}
+
+		iter1 := NewFixedIterator(pathCaveat1)
+		iter2 := NewFixedIterator(pathNoCaveat)
+		iter3 := NewFixedIterator(pathCaveat2)
+
+		intersect := NewIntersection()
+		intersect.addSubIterator(iter1)
+		intersect.addSubIterator(iter2)
+		intersect.addSubIterator(iter3)
+
+		relSeq, err := ctx.Check(intersect, NewObjects("document", "doc1"), NewObject("user", "alice").WithEllipses())
+		require.NoError(err)
+
+		rels, err := CollectAll(relSeq)
+		require.NoError(err)
+
+		require.Len(rels, 1, "Should return one intersected path")
+		require.NotNil(rels[0].Caveat, "Final result should have combined caveat")
+		// The final caveat should be some combination of caveat1 and caveat2 (implementation detail)
```

Comment:
I'm confused on this point - if one of the potential paths exists and doesn't have a caveat, can't you ignore the caveated paths?